### PR TITLE
Archive project with RAMP

### PR DESCRIPTION
**Pull request by vkarthim**
*Added interface for RAMPArchive
*Added menu in Project
*Added logic to "Archive with RAMP (SIL).." menu

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/saymore/37)

### CODE REVIEW

**Comment by gtryus** [comment]
Karthi, I commented in the trello board that I would like the list of contributors to be in the RAMP meta data. Also the audience should normally be set to "Wider audience" for Saymore projects and the kind of data is other.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/saymore/37#-:-LCZg2Wg0YZoMCXVsJGu:b7n3jic)*

**Comment by gtryus** [comment]
For the "List of contributors, this has not worked for Sessions either so we plan to handle this as its own card in trello

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/saymore/37#-:-LCeVuvJBtvbwfzpb2xc:b-u14icn)*

File: src/SayMore/Model/ArchivingHelper.cs [hunk 1]
```diff
@@ -66,6 +67,32 @@ internal static void ArchiveUsingIMDI(IIMDIArchivable element)
...
+				element.ArchiveInfoDetails, element.SetFilesToArchive, element.GetFileDescription)
+			{
+				FileCopyOverride = FileCopySpecialHandler
+			};
+			model.AppSpecificFilenameNormalization = element.CustomFilenameNormalization;
+			model.OverrideDisplayInitialSummary = fileLists => element.DisplayInitialArchiveSummary(fileLists, model);
+			model.HandleNonFatalError = (exception, s) => ErrorReport.NotifyUserOfProblem(exception, s);
+
+			element.InitializeModel(model);
+			element.SetAdditionalMetsData(model);
```
Comment (jasonleenaylor) [inline src/SayMore/Model/ArchivingHelper.cs]: Should this be named SetAdditionalMetaData?

File: src/SayMore/Model/ArchivingHelper.cs [hunk 2]
```diff
@@ -139,6 +166,27 @@ internal static void SetIMDIMetadataToArchive(IIMDIArchivable element, Archiving
...
+			{
+				AddRAMPSession((Session)element, model as RampArchivingDlgViewModel);
+			}
+		}
+
+		internal static void AddRAMPSession(Session saymoreSession, RampArchivingDlgViewModel model)
+		{
+			var contributions = saymoreSession.MetaDataFile.GetValue("contributions", null) as ContributionCollection;
+			if (contributions != null && contributions.Count > 0)
+				model.SetContributors(new ContributionCollection());
```
Comment (jasonleenaylor) [inline src/SayMore/Model/ArchivingHelper.cs]: Why is this passing in an empty ContributionCollection? Do we even need this or does it get handled later with a call to SetAdditionalMetsData?

File: src/SayMore/Model/Project.cs [hunk 3]
```diff
@@ -312,6 +315,56 @@ public void Save()
...
+				// Return total duration of source audio/video recordings.
+				TimeSpan totalDuration = session.GetTotalDurationOfSourceMedia();
+				if (totalDuration.Ticks > 0)
+					model.SetAudioVideoExtent(string.Format("Total Length of Source Recordings: {0}", totalDuration.ToString()));
+
+				//Single session details are enough for "Archive RAMP (SIL)..." from Project menu
+				break;
+			}
+
+			//model.SetSoftwareRequirements("SayMore");
```
[the remark this answers is not in the document]
Comment (vkarthim) [reply src/SayMore/Model/Project.cs]: Ok

File: src/SayMore/Model/ArchivingHelper.cs (hunk 1 above)
Comment (vkarthim) [reply src/SayMore/Model/ArchivingHelper.cs]: I think it defines the logic for adding data to the mets.xml file which involving in RAMP Package.

Comment (jasonleenaylor) [reply src/SayMore/Model/ArchivingHelper.cs]: ah, ok. :)

**Comment by gtryus** [comment]
Reworked the test to mock add files so we can test the files added without reading the zip file. See https://github.com/sillsdev/saymore/pull/39

**Pull request closed without merging**